### ISSUE TITLE
Pyscaffold pytorch extension

### ISSUE THREAD

**Issue by prassanna-ravishankar** [pyscaffold/pyscaffold#286]
## Describe your use-case
Hey, I was quite impressed by the --dsproject extension, and I'm using it heavily for a bunch of projects. I would like to contribute and create a pytorch specific extension which basically combines pyscaffoldext-dsproject  and a commonly used pytorch template such as https://github.com/L1aoXingyu/Deep-Learning-Project-Template.

I've currently started making it on [pyscaffoldext-custom-extension](https://github.com/pyscaffold/pyscaffoldext-custom-extension) along with [pyscaffoldext-dsproject](https://github.com/pyscaffold/pyscaffoldext-dsproject), but I have a couple of questions :- 
1. If I make it, is it possible to get someone to review it?
2. How do I test it locally?
3. Are --dsproject and --custom-extension incompatible? I see some missing files with it.
4. Should I rather clone --dsproject and start from there?

## Describe the solution you would like to see for your use-case
An extension that is adapted *just* for PyTorch.

**Comment by atemysemicolon** [comment]
Also, Can someone help with a snippet to add a pre-designed source/package/module structure ?

**Comment by abravalheri** [comment]
Hello @atemysemicolon, very nice to hear this!

I will try to answer your questions a bit out of order

**1** . I can try to give a review for the code in general, but for datascience stuff maybe @FlorianWilhelm is the best (I don't know how available he is though).

**3/4**. `dsproject` will generate a project that is ready to be used for datascience, while `custom-extension` is super meta and will generate a project that generates projects... That is why it might be a bit confusing. I would consider the following:
   - If your extension changes most of the files generated by `dsproject`, it might be best to clone `dsproject` and change it accordingly
   - If your extension just add a few extra touches/files to the structure generated by `dsproject`, you can try to use `putup --custom-extension` and automatically invoke `dsproject`...
     You can see this approach of automatically invoking other extensions [here](https://github.com/pyscaffold/pyscaffoldext-custom-extension/blob/ae6ff8b1ebd7ffa6be94274d2dd9e1c6e4917b93/src/pyscaffoldext/custom_extension/extension.py#L45).
     Then you can simply add more files using [helps.ensure](https://pyscaffold.org/en/latest/api/pyscaffold.api.html#pyscaffold.api.helpers.ensure) or do small modifications in the existing files using [helpers.modify](https://pyscaffold.org/en/latest/api/pyscaffold.api.html#pyscaffold.api.helpers.modify). Just make sure to [register](https://pyscaffold.org/en/latest/api/pyscaffold.api.html#pyscaffold.api.helpers.register) your actions to run after `dsproject`'s  (you can find the correct name of the actions by running `putup --dsproject --list-actions dummy-project` or with `--pretend`)
     There is some documentation/example for all these methods available in [the website](https://pyscaffold.org/en/latest/extensions.html)

**2**. We rely most of the time in automated tests (so we notice when something we change break the program), and we divide the tests into 2 categories:
  - *regular*: these tests are quicker and just check if the right files where created in the right places with the right content, or that some messages are showed to the user. You can find an example in [test_pre_commit.py](https://github.com/pyscaffold/pyscaffold/blob/7d412e9b5907776ed9c6cd254f9d8eb29b8cac1b/tests/extensions/test_pre_commit.py#L12)
  - *system*: these tests are slower and actually create a project using your extension, so you can even run complex tasks (such as running tests that are generated automatically by the extension). You can find an example in [test_common.py](https://github.com/pyscaffold/pyscaffold/blob/7d412e9b5907776ed9c6cd254f9d8eb29b8cac1b/tests/system/test_common.py#L163)

You can always install the extension in editable mode inside a virtualenv to toy around an experience how it is creating projects:

```bash
# Assuming you are inside your extension project
$ python3 -m venv .venv
$ source .venv/bin/activate
$ pip install -e .
# Then you can use your extension to create dummy projects, just to have a fell how it is working
$ cd /tmp
$ which putup # ==> make sure it is the one inside the venv you have created
$ putup --MY_EXTENSION dummy
```

I hope these comments help!

**Comment by atemysemicolon** [comment]
That last one is what I needed. I'll hack away, and hopefully update this issue with a PR. Still not quite clear on how to bring in a package structure though, but I think with the possibility of testing my own extension, should find it with trial and error. 

Just an example folder structure
- `project_root`
    - `src`
        - `my_package`
            - `libs`
            - `network`
            - `tools`
            - `libs`


Basically, where can I find this dict mentioned here : https://pyscaffold.org/en/latest/extensions.html#project-structure-representation in `extension.py`?

**Comment by atemysemicolon** [comment]
On another note, I've renamed some stuff and tried hacking away. However, I get the following issue :- 

```
/home/atemysemicolon/.pyenv/versions/torch_scaffold/bin/python /home/atemysemicolon/.pyenv/versions/torch_scaffold/bin/putup --pytorch dummy
Traceback (most recent call last):
  File "/home/atemysemicolon/.pyenv/versions/torch_scaffold/bin/putup", line 11, in <module>
    sys.exit(run())
  File "/home/atemysemicolon/.pyenv/versions/torch_scaffold/lib/python3.6/site-packages/pyscaffold/shell.py", line 72, in func_wrapper
    func(*args, **kwargs)
  File "/home/atemysemicolon/.pyenv/versions/torch_scaffold/lib/python3.6/site-packages/pyscaffold/utils.py", line 145, in func_wrapper
    func(*args, **kwargs)
  File "/home/atemysemicolon/.pyenv/versions/torch_scaffold/lib/python3.6/site-packages/pyscaffold/cli.py", line 232, in run
    main(sys.argv[1:])
  File "/home/atemysemicolon/.pyenv/versions/torch_scaffold/lib/python3.6/site-packages/pyscaffold/cli.py", line 223, in main
    opts = parse_args(args)
  File "/home/atemysemicolon/.pyenv/versions/torch_scaffold/lib/python3.6/site-packages/pyscaffold/cli.py", line 137, in parse_args
    in iter_entry_points('pyscaffold.cli')]
  File "/home/atemysemicolon/.pyenv/versions/torch_scaffold/lib/python3.6/site-packages/pyscaffold/cli.py", line 136, in <listcomp>
    cli_extensions = [extension.load()(extension.name) for extension
  File "/home/atemysemicolon/.pyenv/versions/torch_scaffold/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2324, in load
    return self.resolve()
  File "/home/atemysemicolon/.pyenv/versions/torch_scaffold/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2330, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
ModuleNotFoundError: No module named 'pyscaffoldext.dsproject'

Process finished with exit code 1
```

Strange because I can see it :- 
```
pip freeze | grep markdown
pyscaffoldext-markdown==0.3.2
```

and Strange, because it actually exists in site-packages
```
➜  markdown git:(master) pwd
/home/atemysemicolon/.pyenv/versions/torch_scaffold/lib/python3.6/site-packages/pyscaffoldext/markdown

➜  markdown git:(master) ls -all
total 28
drwxrwxr-x 4 atemysemicolon atemysemicolon 4096 Jul  9 23:55 .
drwxrwxr-x 4 atemysemicolon atemysemicolon 4096 Jul  9 23:55 ..
-rw-rw-r-- 1 atemysemicolon atemysemicolon 4336 Jul  9 23:55 extension.py
-rw-rw-r-- 1 atemysemicolon atemysemicolon  379 Jul  9 23:55 __init__.py
drwxrwxr-x 2 atemysemicolon atemysemicolon 4096 Jul  9 23:55 __pycache__
drwxrwxr-x 3 atemysemicolon atemysemicolon 4096 Jul  9 23:55 templates
```

```
(torch_scaffold) ➜  pyscaffoldext-pytorch git:(master) ✗ pyenv which python
/home/atemysemicolon/.pyenv/versions/torch_scaffold/bin/python

```

**Comment by atemysemicolon** [comment]
Sorry for bothering. Got it!

```

class PytorchProject(Extension):
    """Template for data-science projects
    """

    def augment_cli(self, parser):
        """Augments the command-line interface parser

        A command line argument ``--FLAG`` where FLAG=``self.name`` is added
        which appends ``self.activate`` to the list of extensions. As help
        text the docstring of the extension class is used.
        In most cases this method does not need to be overwritten.

        Args:
            parser: current parser object
        """
        help = self.__doc__[0].lower() + self.__doc__[1:]
        print(parser)

        parser.add_argument(
            self.flag, help=help, nargs=0, dest="extensions", action=IncludeExtensions
        )
        return self

    def activate(self, actions):
        # First action a copy paste of dsproject actually :)
        actions = self.register(actions, add_pytorch_project, after="define_structure")
        actions = self.register(actions, add_pytorch_structure, after="add_pytorch_project")
        actions = self.register(actions, replace_readme, after="add_pytorch_structure")
        return actions


def add_pytorch_structure(struct, opts):
    """Adds basic module for custom extension

    Args:
        struct (dict): project representation as (possibly) nested
            :obj:`dict`.
        opts (dict): given options, see :obj:`create_project` for
            an extensive list.

    Returns:
        struct, opts: updated project representation and options
    """
    if "namespace" in opts:
        path = [opts["project"], "src", opts["namespace"], opts["package"], "lib",
                "nn_function.py"]
    else:
        path = [opts["project"], "src", opts["package"], "lib",
                "nn_function.py"]
    train_model_py = templates.train_model_py(opts)
    struct = helpers.ensure(struct, path, train_model_py, helpers.NO_OVERWRITE)
    return struct, opts
```

**Comment by abravalheri** [comment]
@atemysemicolon it is looking nice, good to see you are getting used to the API.

**Comment by FlorianWilhelm** [comment]
@atemysemicolon, I am looking forward to your extension. Let us know if you want to host and maintain your extension under the [PyScaffold Github Organisation](https://github.com/pyscaffold).

**Comment by atemysemicolon** [comment]
will ping once i'm done.

**Comment by atemysemicolon** [comment]
@FlorianWilhelm  You can see my progress here : https://github.com/atemysemicolon/pyscaffoldext-pytorch/tree/feature/pytorch-extension. The scaffolding works, there's a ton of TODOs here and there though.

Started off with your --dsproject extension

**Comment by atemysemicolon** [comment]
@abravalheri , @FlorianWilhelm  any way I can give a custom `.pre-commit-config.yaml` file via a template? I am not sure if I am doing something wrong, but it seems like the pre-commit's `isort` isn't picking up configurations in `.isort.cfg`

**Comment by BAILOOL** [comment]
Looking forward to seeing this happen. This will be a great contribution to the Pytorch community.

**Comment by abravalheri** [comment]
> @abravalheri , @FlorianWilhelm any way I can give a custom `.pre-commit-config.yaml` file via a template? I am not sure if I am doing something wrong, but it seems like the pre-commit's `isort` isn't picking up configurations in `.isort.cfg`

Hi @atemysemicolon, yes you can if you want... you have to replace the file in the dict structure... I wouldn't advise the complete replacement though... that sounds very dramatic, but appending more hooks to the end of the file should be completely fine.

Regarding the isort stuff, sometimes updating the installation might help.

**Comment by FlorianWilhelm** [comment]
Hi @atemysemicolon, how is pyscaffoldext-pytorch coming along? If you have a first version, we can also create the corresponding repository within the pyscaffold organisation. It seems that you sparked much interest and I would also love to see a pytorch extension :-)

**Comment by FlorianWilhelm** [comment]
Hi @atemysemicolon, how is your PyTorch extension coming along? And will you include fastai, ignite or lightning?

**Comment by FlorianWilhelm** [comment]
[PING] It's been more than a year since we heard from @prassanna-ravishankar. Also, the last commit [in his repo](https://github.com/prassanna-ravishankar/pyscaffoldext-pytorch/tree/feature/pytorch-extension) is older than a year. The core PyScaffold contributors would still love to see this extension happen and also @BAILOOL showed interest. Our offer still stands to make this an official PyScaffold extension when it is finished.

If there is no update within the next month, i.e. November 2021, my suggestion would be to close this issue for the time being.

**Comment by FlorianWilhelm** [comment]
Closing this for now. Please feel free to reopen if anyone else wants to work on this :-)
